plutus-core/src/Language/PlutusCore/Core/Instance/Eq.hs

```diff
@@ -50,6 +53,146 @@ instance
+does not handle this case correctly.
+
+Note however that we do want to handle shadowing like that when a lambda that binds a variable
+shadowing an existing one is not amongst the leading lambdas, i.e. it's either applied in the body
+of the n-ary lambda term or an argument there. For example, when performing this equality
+check:
+
+    f (\x -> x) == (\x -> f (\x -> x) x)
+
+we do want the first @x@ bound on the RHS to be shadowed by the second one.
+
+What we can do here is remove all the leading lambdas in a single pass and ensure that the
+variables they bind are all distinct (as @\x x -> f@ can never be eta-contracted into @f@
+regardless of what @f@ is).
+
+But with the inefficient solution we don't need to check that all bound variables are distinct.
+Instead, we can compute the set of free variables of the fully eta-contracted function and
+recurse on the list of bound variables doing this:
+
+1. check that the variable is not amongst free variables
+2. add it to the set of free variables
+
+So for @\x x -> f x x@ we first eta-contract it fully to @f@, then check that the first @x@ is not
+free in @f@ (which it's clearly not, given that @f@ is a type variable) and then check that the
+second @x@ is not free in @f x@, which it is, so we fail here.
+
+Regardless of which solution is chosen, it wouldn't be correct to use such failing eta-contraction
+for both the sides of an equation first and only then check their equality, because then we
+wouldn't be able to eta-contract the RHS of this equality check that is meant to succeed:
+
+    (\x -> f) == (\x x -> f x)
+
+However the case where both the sides start with a lambda is already correctly handled by the main
+machinery, so when we do eta-contraction, we can safely assume that the other side of the equation
+does not have any leading lambdas.
+-}
+
+-- See Note [Eta-equality].
+fullyEtaContract
+    :: forall ren tyname uni ann. HasUnique tyname TypeUnique
+    => Type tyname uni ann -> EqRenameT ren Maybe (Type tyname uni ann)
+fullyEtaContract = stripTyLams [] where
+    stripTyLams :: [TypeUnique] -> Type tyname uni ann -> EqRenameT ren Maybe (Type tyname uni ann)
+    stripTyLams binds (TyLam _ name _ body') = stripTyLams (view unique name : binds) body'
+    stripTyLams binds body                   = stripTyApps binds body
+
+    stripTyApps :: [TypeUnique] -> Type tyname uni ann -> EqRenameT ren Maybe (Type tyname uni ann)
+    stripTyApps binds0 = go binds0 where
+        go :: [TypeUnique] -> Type tyname uni ann -> EqRenameT ren Maybe (Type tyname uni ann)
+        go [] fun = fun <$ checkScoping (ftuTy fun) binds0
+        go (bind1 : binds) (TyApp _ fun' (TyVar _ name2))
+            | bind1 == name2 ^. unique = go binds fun'
+        go _ _ = empty
+
+    checkScoping :: Set.Set TypeUnique -> [TypeUnique] -> EqRenameT ren Maybe ()
+    checkScoping _   []             = pure ()
+    checkScoping ftu (bind : binds) = do
+        guard $ bind `Set.notMember` ftu
```

Comment:
Irrelevant, but sometimes I miss the Java set-insertion functions which return a boolean telling you whether the item you inserted was already in the set or not. Surprisingly useful.

Reply:
That would be nice to have indeed.

Reply:
I think the semantics os `java.set.add` method vs `Data.Set.insert` are slightly different. The java one will not override if the an object is already in the set (using object.equals) whereas the haskell one actually *always* overrides the element (using the Eq instance) . See <https://docs.oracle.com/javase/7/docs/api/java/util/Set.html#add(E)> and <https://hackage.haskell.org/package/containers-0.6.4.1/docs/Data-Set.html#v:insert>  . So in Haskell terms, the `Data.Set.insert` always returns True ;)

Reply:
That makes me think, is this guard here redundant, computationally-wise? (assuming the Eq instance is kind of faithful)

Reply:
@bezirg 

> This makes me think, is this guard here redundant, computationally-wise? (assuming the Eq instance is kind of faithful)

Note that we first check and only then insert, not the other way around. If that guard is commented out, then tests fails with something like "`\x -> f x x` is not supposed to be equal to `f x`" (just checked to be sure).